src/mlpack/core/metrics/bleu_score_impl.hpp

```diff
@@ -0,0 +1,154 @@
+/**
+ * @file core/metrics/bleu_score_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of BLEUScore class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_METRICS_BLEU_SCORE_IMPLHPP
+#define MLPACK_CORE_METRICS_BLEU_SCORE_IMPLHPP
+
+// In case it hasn't been included.
+#include "bleu_score.hpp"
+
+namespace mlpack {
+namespace metric {
+
+template <typename ElemType, typename PrecisionType>
+BLEU<ElemType, PrecisionType>::BLEU(const size_t maxOrder) : maxOrder(maxOrder)
+{
+  // Nothing to do here.
+}
+
+template <typename ElemType, typename PrecisionType>
+template <typename WordVector>
+std::map<WordVector, size_t> BLEU<ElemType, PrecisionType>::GetNGrams(
+  const WordVector& segment)
+{
+  std::map<WordVector, size_t> ngramsCount;
+  for (size_t order = 1; order < maxOrder + 1; ++order)
+  {
+    for (size_t i = 0; i < segment.size() - order + 1; ++i)
+    {
+      WordVector seq = WordVector(segment.begin() + i,
+                                  segment.begin() + i + order);
+      ngramsCount[seq]++;
+    }
+  }
+  return ngramsCount;
+}
+
+template <typename ElemType, typename PrecisionType>
+template <typename ReferenceCorpusType, typename TranslationCorpusType>
+ElemType BLEU<ElemType, PrecisionType>::Evaluate(
+  const ReferenceCorpusType& referenceCorpus,
+  const TranslationCorpusType& translationCorpus,
+  const bool smooth)
+{
+  typedef typename TranslationCorpusType::value_type WordVector;
+  std::vector<size_t> matchesByOrder(maxOrder, 0);
+  std::vector<size_t> possibleMatchesByOrder(maxOrder, 0);
+  referenceLength = 0, translationLength = 0;
+
+  auto refIt = referenceCorpus.cbegin();
+  auto trIt = translationCorpus.cbegin();
+  for(; refIt != referenceCorpus.cend(), trIt != translationCorpus.cend();
+      ++refIt, ++trIt)
+  {
+    size_t min = std::numeric_limits<size_t>::max();
+    for (auto t: *refIt)
+    {
+      if (min > t.size())
+      {
+        min = t.size();
+      }
+    }
+    referenceLength += min;
+    translationLength += trIt->size();
+
+    std::map<WordVector, size_t> mergedRefNGramCounts;
+    for (auto t: *refIt)
+    {
+      const std::map<WordVector, size_t> ngrams = GetNGrams(t);
+      for (auto it = ngrams.cbegin(); it != ngrams.cend(); ++it)
+      {
+        if (!mergedRefNGramCounts[it->first])
+          mergedRefNGramCounts[it->first] = it->second;
+        else
+          mergedRefNGramCounts[it->first]
+              = std::max(mergedRefNGramCounts[it->first], it->second);
+      }
+    }
+
+    std::map<WordVector, size_t> translationNGramCounts = GetNGrams(*trIt);
+    std::map<WordVector, size_t> overlap;
```

Comment:
We can use `boost::string_view` instead of `std::string` here and many other place too

Reply:
I've used 
```cpp
typedef typename TranslationCorpusType::value_type WordVector;
``` 
statement in line number 52. I think it's up to the user what to be used.

Reply:
It is ok to give user a choice, but here overlap variable doesn't have any relation with the user, I mean it is not exposed to user

`map<<vector<string>>, size_t> overlap` , it would be unnecessary copy of whole string when inserting the value in overlap map

That's my POV, may be you can wait for someone to give a better insight

Reply:
But I think it will be better to add tests for both types :)

Reply:
>it would be unnecessary copy of whole string when inserting the value in overlap map

Ok I will see this. Thanks for clarifying.